matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/DefaultCryptoService.kt

```diff
@@ -907,27 +813,168 @@ internal class DefaultCryptoService @Inject constructor(
         }
     }
 
-    /**
-     * Upload my user's device keys.
-     */
-    private suspend fun uploadDeviceKeys() {
-        if (cryptoStore.areDeviceKeysUploaded()) {
-            Timber.d("Keys already uploaded, nothing to do")
-            return
+    private fun notifyRoomKeyReceived(
+            roomId: String,
+            sessionId: String,
+    ) {
+        // The sender key is actually unused since it's unimportant for megolm
+        // Our events don't contain the info so pass an empty string until we
+        // change the listener definition
+        val senderKey = ""
+
+        newSessionListeners.forEach {
+            try {
+                it.onNewSession(roomId, senderKey, sessionId)
+            } catch (e: Throwable) {
+            }
         }
-        // Prepare the device keys data to send
-        // Sign it
-        val canonicalJson = JsonCanonicalizer.getCanonicalJson(Map::class.java, getMyDevice().signalableJSONDictionary())
-        var rest = getMyDevice().toRest()
+    }
 
-        rest = rest.copy(
-                signatures = objectSigner.signObject(canonicalJson)
-        )
+    suspend fun receiveSyncChanges(
+            toDevice: ToDeviceSyncResponse?,
+            deviceChanges: DeviceListResponse?,
+            keyCounts: DeviceOneTimeKeysCountSyncResponse?) {
+        // Decrypt and handle our to-device events
+        val toDeviceEvents = this.olmMachine!!.receiveSyncChanges(toDevice, deviceChanges, keyCounts)
+
+        // Notify the our listeners about room keys so decryption is retried.
+        if (toDeviceEvents.events != null) {
+            toDeviceEvents.events.forEach { event ->
+                when (event.type) {
+                    EventType.ROOM_KEY           -> {
+                        val content = event.getClearContent().toModel<RoomKeyContent>() ?: return@forEach
+
+                        val roomId = content.sessionId ?: return@forEach
+                        val sessionId = content.sessionId
+
+                        notifyRoomKeyReceived(roomId, sessionId)
+                    }
+                    EventType.FORWARDED_ROOM_KEY -> {
+                        val content = event.getClearContent().toModel<ForwardedRoomKeyContent>() ?: return@forEach
+
+                        val roomId = content.sessionId ?: return@forEach
+                        val sessionId = content.sessionId
+
+                        notifyRoomKeyReceived(roomId, sessionId)
+                    }
+                    else                         -> {
+                        this.verificationService?.onEvent(event)
+                    }
+                }
+            }
+        }
+    }
+
+    private suspend fun preshareRoomKey(roomId: String, roomMembers: List<String>) {
+        keyClaimLock.withLock {
```

Comment:
Isn't this locking a bit too much? Like it will keep lock while doing server request?

Reply:
Why is that too much? Is it a problem to keep the lock while sending the request out? 

The point is to not generate and send out the same request twice. The state machine will generate the same request multiple times unless you tell it that the given request was sent out.